### PR TITLE
feat: Reworked Buffer and Flusher Interfaces and implementations.

### PR DESCRIPTION
**Pull request by cpheps**
## Description of Changes
- Reworked the buffer interface
- Rewrote memory and disk buffers
- Refactored how flusher works
- Updated operators that use buffer and flusher

## **Please check that the PR fulfills these requirements**
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)
- [x] CI passes

### CODE REVIEW

File: operator/builtin/output/forward/forward.go [hunk 1]
```diff
@@ -101,7 +103,34 @@ func (f *ForwardOutput) Stop() error {
...
+	entries, err := f.buffer.Close()
+	if err != nil {
+		f.Error("Failed to retrieve entries")
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	err = f.sendEntries(ctx, entries)
```
Comment (BinaryFissionGames) [inline operator/builtin/output/forward/forward.go:116]: Looks like we aren't protecting against len(entries) == 0, so it looks like we'll send a request with 0 entries?

File: operator/builtin/output/elastic/elastic.go [hunk 2]
```diff
@@ -125,7 +126,42 @@ func (e *ElasticOutput) Stop() error {
...
+	entries, err := e.buffer.Close()
+	if err != nil {
+		e.Error("Failed to retreive entries")
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	err = e.sendEntries(ctx, entries)
```
Comment (BinaryFissionGames) [inline operator/builtin/output/elastic/elastic.go:139]: Looks like we aren't protecting against len(entries) == 0, so it looks like we'll send a request with 0 entries?

File: operator/buffer/disk_test.go [hunk 3]
```diff
@@ -4,347 +4,676 @@ import (
...
-		err = b2.Open(dir, false)
-		require.NoError(t, err)
-		readN(t, b2, 10, 10)
-		err = b2.Close()
-		require.NoError(t, err)
-	})
+func TestDiskBufferConcurrency(t *testing.T) {
+	var (
+		numEntries = 10000
+		timeout    = 15 * time.Second
```
Comment (BinaryFissionGames) [inline operator/buffer/disk_test.go:456]: Looks like this timeout got reverted

Comment (cpheps) [reply operator/buffer/disk_test.go:456]: interesting... I'll fix up

**Review by armstrmi — APPROVED** [review]

**Review by BinaryFissionGames — APPROVED** [review]

**Pull request merged**